File: 4. linkedList/拆分链表.java
/*
给定一个链表，需要把链表从中间拆分成长度相等的两半（如果链表长度为奇数，那么拆分之后，前半部分长度更长一点）。

输入：[1->2->3->4->5]

输出：[1->2->3, 4->5]

【分析】我们需要分为 2 步：

1. 找到链表的中间结点

2. 从中间结点把链表分为两半
 */
class Solution{
    public ListNode findMiddleNode(ListNode head) {
        ListNode dummy = new ListNode();
        dummy.next = head;
        ListNode s1 = head;
        ListNode s2 = head;
        ListNode pre = dummy;
        while (s2 != null && s2.next != null) {
            s2 = s2.next.next;
            pre = s1;
            s1 = s1.next;
        }
        return s2 == null ? pre:s1;
    }

    public ListNode[] split(ListNode head) {
        ListNode mid = findMiddleNode(head);
        ListNode back = mid.next;
        mid.next = null;
        return new ListNode[]{head, back};
    }
}

File: 04. linkedList/拆分链表.py
"""
【题目】给定一个链表，需要把链表从中间拆分成长度相等的两半（如果链表长度为奇数，那么拆分之后，前半部分长度更长一点）。

输入：[1->2->3->4->5]

输出：[1->2->3, 4->5]

【分析】我们需要分为 2 步：

找到链表的中间结点

从中间结点把链表分为两半
"""
class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution(object):
    def splitList(self, head):
        mid = self.splitList(head)
        back = mid.next
        mid.next = None

        return head, back

    def find_middle(self, head):
        front = head
        back = head
        pre = back
        while front is not None and front.next:
            pre = back
            front = front.next.next
            back = back.next
        return back if front else pre
